flare/ase/calculator.py

```diff
@@ -11,139 +11,129 @@
 from flare.env import AtomicEnvironment
 from flare.struc import Structure
 from flare.mgp import MappedGaussianProcess
-from flare.predict import predict_on_structure_par_en, predict_on_structure_en
+from flare.predict import predict_on_structure_par_en, \
+    predict_on_structure_en, predict_on_structure_efs, \
+    predict_on_structure_efs_par
 from ase.calculators.calculator import Calculator
 
+
 class FLARE_Calculator(Calculator):
     """
     Build FLARE as an ASE Calculator, which is compatible with ASE Atoms and
-    Molecular Dynamics. 
+    Molecular Dynamics.
     Args:
         gp_model (GaussianProcess): FLARE's Gaussian process object
-        mgp_model (MappedGaussianProcess): FLARE's Mapped Gaussian Process object. 
-            `None` by default. MGP will only be used if `use_mapping` is set to True
-        par (Bool): set to `True` if parallelize the prediction. `False` by default.
-        use_mapping (Bool): set to `True` if use MGP for prediction. `False` by default.
+        mgp_model (MappedGaussianProcess): FLARE's Mapped Gaussian Process
+            object. `None` by default. MGP will only be used if `use_mapping`
+            is set to True.
+        par (Bool): set to `True` if parallelize the prediction. `False` by
+            default.
+        use_mapping (Bool): set to `True` if use MGP for prediction. `False`
+            by default.
     """
 
     def __init__(self, gp_model, mgp_model=None, par=False, use_mapping=False):
-        super().__init__() # all set to default values, TODO: change
+        super().__init__()  # all set to default values, TODO: change
         self.mgp_model = mgp_model
         self.gp_model = gp_model
         self.use_mapping = use_mapping
         self.par = par
         self.results = {}
 
-    def get_property(self, name, atoms=None, allow_calculation=True):
+    def get_property(self, name, atoms=None, allow_calculation=True,
+                     structure=None):
         if name not in self.results.keys():
             if not allow_calculation:
                 return None
-            self.calculate(atoms)
+            self.calculate(atoms, structure)
```

Comment:
The interface as implemented is consistent with the general ASE calculator template described [here](https://wiki.fysik.dtu.dk/ase/ase/calculators/calculators.html#calculator-interface). Note that our get_forces, get_potential_energy, and get_stress methods take in an atoms object only; the structure object is only introduced as an optional input to the auxiliary get_property method.

Reply:
Actually all the `get_forces`, `get_stress` etc. calls `get_property`, the reason `atoms.get_forces()` works in `otf` is that the line 144 (ase/otf.py)
```
        self.atoms.calc.calculate(self.atoms, self.structure)
```
called the `calculate` function, so there is a dict `self.atoms.calc.results` which has been written the FLARE calculated `forces`, `energy` etc. Then when the `atoms.get_forces` is called, it will just get the value from the `results` dict without re-calculating of GP, so it's ok at this time if `structure=None`.

This means that it only works in the `otf` context. If we want to create a FLARE_Calculator without OTF, and try to use `get_forces`, the structure has to be input, because at this time `results` dict is empty, and `calculate` function needs calling.

But I notice that `structure` and ASE Atoms are very similar, so probably a better way to do this is just in `ASE_OTF` let
```
    self.structure = self.atoms
```
since they have a lot of common attributes like positions, cell etc. so there should only be some minor changes. may have to think about this. I will try to deal with this in another PR

Reply:
Hmm I actually think get_forces, get_stress etc. should still work without giving a structure. If a structure isn't given, the calculate method creates a structure based on the atoms object (see lines 69-73 of ase/calculator). But I agree this isn't a perfect solution, and it would be better to completely eliminate structure from the ase calculator. This would require adding some extra calculator properties (energy, force, stress uncertainties) and then writing them to the structure somewhere else, probably in a method of ase/otf.